components/heap/third_party/dartino/cmpctmalloc.c

```diff
@@ -1291,24 +1397,121 @@ IRAM_ATTR size_t cmpct_get_allocated_size_impl(cmpct_heap_t *heap, void *p)
+    if (new_pages == old_pages) return p;
+    size_t page = page_number(heap, p);
+    if (new_pages > old_pages) {
+        /* Growing a page allocation. */
+        for (size_t i = page + old_pages; i < page + new_pages; i++) {
+            if (heap->pages[i].status != PAGE_FREE) {
+                /* Failed to grow only forwards - can we grow backwards? */
+                if (i < new_pages) return NULL;
+                size_t first_page = i - new_pages;
+                for (size_t j = first_page; j < page; j++) {
+                    if (heap->pages[j].status != PAGE_FREE) {
+                        return NULL;
+                    }
+                }
+                return page_grow_allocation(heap, p, page, old_pages, first_page, new_pages);
+            }
+        }
+        return page_grow_allocation(heap, p, page, old_pages, page, new_pages);
+    } else {
+        /* Shrinking a page allocation. */
+        for (size_t i = page + new_pages; i < page + old_pages; i++) {
+            heap->pages[i].status = PAGE_FREE;
+        }
+        if (heap->pages[page + old_pages].status != PAGE_FREE) {
+            heap->free_blocks++;
+        }
+        heap->remaining += (old_pages - new_pages) << PAGE_SIZE_SHIFT;
+        return p;
+    }
+}
+
+/* This realloc implementation always first tries to create a new allocation
+   and copy the data.  There are so few chances to defragment a malloc heap
+   that we will try to move the data when we can.
+
+   If this fails we may just return the original area if the size is close
+   enough.
+
+   For large (page-based) failing reallocations we attempt a new allocation
+   that overlaps with the old one.  */
 IRAM_ATTR void *cmpct_realloc_impl(cmpct_heap_t *heap, void *p, size_t size)
 {
     if (!size) {
         cmpct_free_impl(heap, p);
+        /* C++ "new" does not like null responses for zero-length allocations,
+           but it never calls realloc so we can get away with it here.  */
         return NULL;
     }
     void *new_allocation = cmpct_malloc_impl(heap, size);
-    if (!p || !new_allocation) return new_allocation;
+    if (!p) return new_allocation;
     size_t old_size = cmpct_get_allocated_size_impl(heap, p);
-    memcpy(new_allocation, p, MIN(old_size, size));
-    cmpct_free_impl(heap, p);
-    return new_allocation;
+    if (new_allocation) {
+        memcpy(new_allocation, p, MIN(old_size, size));
+        cmpct_free_impl(heap, p);
+        return new_allocation;
+    }
+    if (is_page_allocated(heap, p)) {
+        lock(heap);
+        void *result = realloc_page_allocation(heap, p, size, old_size);
+        unlock(heap);
+        return result;
+    } else {
+        /* We know the old area was not page allocated, which puts an
+           upper bound on how large it could be.  Therefore it can't
+           overflow the calculation below.  */
+        if (size <= old_size && size + sizeof(free_t) > (old_size * 4) / 3) {
```

Comment:
Do you have this cutoff elsewhere?

Reply:
No, it's a more relaxed cutoff than we have elsewhere.